src/pypromice/process/L1toL2.py

```diff
@@ -103,6 +118,28 @@ def toL2(
         lat = ds['gps_lat'].mean()
         lon = ds['gps_lon'].mean()
 
+    # smoothing tilt
```

Comment:
It looks like you are doing 72 hour standard deviation filter with different thresholds for tilt and rotation. And in addition a subsequent smoothing on the rotation data. Correct?

I find this part difficult to read. Consider splitting out the functionality to a helper function for readability testability.
I've asked ChatGPT for a suggestion 😄 :

```python
import xarray as xr
import pandas as pd

def apply_std_threshold_filter(data: xr.DataArray, resample_period: str, rolling_window_size: int,
                               std_threshold: float, time_dim: str = 'time') -> xr.DataArray:
    """Apply a standard deviation threshold filter to time series data."""
    # Convert to pandas Series, resample, and calculate rolling standard deviation
    series: pd.Series = data.to_series()
    medians: pd.Series = series.resample(resample_period).median()
    std_dev: pd.Series = medians.rolling(window=rolling_window_size, center=True, min_periods=2).std()
    reindexed_std: np.ndarray = std_dev.reindex(data[time_dim], method='bfill').values

    # Filter data based on standard deviation threshold
    filtered_data: xr.DataArray = data.where(reindexed_std < std_threshold)
    return filtered_data.ffill(dim=time_dim).bfill(dim=time_dim)

def process_tilt(ds: xr.Dataset, variables: list[str], std_threshold: float = 0.2) -> None:
    """Process tilt data variables using a standard deviation threshold."""
    for variable in variables:
        ds[variable] = apply_std_threshold_filter(ds[variable], 'H', 3*24, std_threshold)

def process_rotation(ds: xr.Dataset, variable: str = 'rot', std_threshold: float = 4,
                     smoother_window_size: int = 14) -> None:
    """Process rotation data with a harsher treatment and additional smoothing."""
    # Apply harsher std threshold filtering
    filtered_data: xr.DataArray = apply_std_threshold_filter(ds[variable], 'H', 3*24, std_threshold)

    # Additional smoothing by resampling to daily and applying median rolling
    daily_medians: pd.Series = filtered_data.to_series().resample('D').median()
    smoothed_daily: pd.Series = daily_medians.rolling(window=smoother_window_size, center=True, min_periods=2).median()
    reindexed_smoothed: np.ndarray = smoothed_daily.reindex(ds.time, method='bfill').values

    # Update the dataset
    ds[variable] = ('time', reindexed_smoothed)

# Example usage:
ds = xr.Dataset({
    'tilt_x': ('time', pd.Series(range(100))),
    'tilt_y': ('time', pd.Series(range(100))),
    'rot': ('time', pd.Series(range(100)))
})
ds['time'] = pd.date_range(start='2020-01-01', periods=100, freq='T')

process_tilt(ds, ['tilt_x', 'tilt_y'])
process_rotation(ds, 'rot')

```

Reply:
I guess it is a matter of taste.

I don't think a two-line function like `process_tilt` make sense.
I also think that declaring variable type at each operation (e.g. `daily_medians: pd.Series = ...`) decreases readability.

I have moved these lines into two seprate functions:

```python
    ds['tilt_x'] = smoothTilt(ds['tilt_x'])
    ds['tilt_y'] = smoothTilt(ds['tilt_y'])
    ds['rot'] = smoothRot(ds['rot'])

...


def smoothTilt(da: xr.DataArray, threshold=0.2):
    '''Smooth the station tilt

    Parameters
    ----------
    da : xarray.DataArray
        either X or Y tilt inclinometer measurements
    threshold : float
        threshold used in a standrad.-deviation based filter

    Returns
    -------
    xarray.DataArray
        either X or Y smoothed tilt inclinometer measurements
    '''
    # we caluclate the moving standard deviation over a 3-day sliding window
    # hourly resampling is necessary to make sure the same threshold can be used 
    # for 10 min and hourly data
    moving_std_gap_filled = da.to_series().resample('H').median().rolling(
                    3*24, center=True, min_periods=2
                    ).std().reindex(da.time, method='bfill').values
    # we select the good timestamps and gapfill assuming that
    # - when tilt goes missing the last available value is used
    # - when tilt is not available for the very first time steps, the first
    #   good value is used for backfill
    return da.where(
                moving_std_gap_filled < threshold
                ).ffill(dim='time').bfill(dim='time')


def smoothRot(da: xr.DataArray, threshold=4):
    '''Smooth the station rotation

    Parameters
    ----------
    da : xarray.DataArray
        rotation measurements from inclinometer
    threshold : float
        threshold used in a standrad-deviation based filter

    Returns
    -------
    xarray.DataArray
        smoothed rotation measurements from inclinometer
    '''
    moving_std_gap_filled = da.to_series().resample('H').median().rolling(
                    3*24, center=True, min_periods=2
                    ).std().reindex(da.time, method='bfill').values
    # same as for tilt with, in addition:
    #     - a resampling to daily values
    #     - a two week median smoothing
    #     - a resampling from these daily values to the original temporal resolution
    return ('time', (da.where(moving_std_gap_filled <4).ffill(dim='time')
            .to_series().resample('D').median()
            .rolling(7*2,center=True,min_periods=2).median()
            .reindex(da.time, method='bfill').values
            ))


```

Is that better?